### PR TITLE
Restore old initial batch distribution logic in LoadScheduling

### PR DESCRIPTION
**Pull request by amezin**
pytest orders tests for optimal sequential execution - i. e. avoiding
unnecessary setup and teardown of fixtures. So executing tests in consecutive
chunks is important for optimal performance.

Commit 09d79ace356ebf14d221cecd3fd8002d075a2034 optimized test distribution for
the corner case, when the number of tests is less than 2 * number of nodes.
At the same time, it made initial test distribution worse for all other cases.
If some tests use some fixture, and these tests fit into the initial batch,
the fixture will be created min(n_tests, n_workers) times, no matter how many
other tests there are. With the old algorithm (before
09d79ace356ebf14d221cecd3fd8002d075a2034), if there are enough tests not using
the fixture, the fixture was created only once.

So restore the old behavior for typical cases where the number of tests is
much greater than the number of workers (or, strictly speaking, when there
are at least 2 tests for every node).

In my test suite, where fixtures create Docker containers, this change reduces
total run time by 10-15%.

This is a partial revert of commit 09d79ace356ebf14d221cecd3fd8002d075a2034

### CODE REVIEW

**Comment by amezin** [comment]
There will be a failing test because of changed scheduling, see https://github.com/pytest-dev/pytest-xdist/pull/813

File: testing/test_dsession.py [hunk 1]
```diff
@@ -115,18 +115,18 @@ def test_schedule_batch_size(self, pytester: pytest.Pytester) -> None:
         # assert not sched.tests_finished
         sent1 = node1.sent
         sent2 = node2.sent
-        assert sent1 == [0, 2]
-        assert sent2 == [1, 3]
+        assert sent1 == [0, 1]
```
Comment (nicoddemus) [inline testing/test_dsession.py:118]: This test was updated to match the new distribution method, could you please add a new test which checks that we are using round robin for smaller collections?

File: src/xdist/scheduler/load.py [hunk 2]
```diff
@@ -248,13 +248,19 @@ def schedule(self):
...
         # shutdown signals and get all nodes working.
-        initial_batch = max(len(self.pending) // 4, 2 * len(self.nodes))
-
-        # distribute tests round-robin up to the batch size
-        # (or until we run out)
-        nodes = cycle(self.nodes)
-        for i in range(initial_batch):
-            self._send_tests(next(nodes), 1)
+        if len(self.pending) < 2 * len(self.nodes):
+            # distribute tests round-robin
```
Comment (nicoddemus) [inline src/xdist/scheduler/load.py:252]: Can you please elaborate the comment to explain why we use round-robin for small collections?

File: src/xdist/scheduler/load.py [hunk 3]
```diff
@@ -248,13 +248,19 @@ def schedule(self):
...
+            nodes = cycle(self.nodes)
+            for i in range(len(self.pending)):
+                self._send_tests(next(nodes), 1)
+        else:
+            # how many items per node do we have about?
```
Comment (nicoddemus) [inline src/xdist/scheduler/load.py:257]: Can you please improve the comment explaining why we fall back to large batches split evenly for the initial batch?

File: changelog/812.feature [hunk 4]
```diff
@@ -0,0 +1 @@
+Partially restore old initial batch distribution algorithm in ``LoadScheduling``.
```
Comment (nicoddemus) [inline changelog/812.feature:1]: Can you please elaborate this? Something like what you used in the PR description would be great. 👍

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @amezin!

Appreciate the PR, please take a look at my comments. 👍

File: testing/test_dsession.py (hunk 1 above)
Comment (amezin) [reply testing/test_dsession.py:118]: There is already such test, called `test_schedule_fewer_than_two_tests_per_node`. Also, the update of this test is simply part of the revert too - it was modified in 09d79ace356ebf14d221cecd3fd8002d075a2034, and I'm just undoing that change.

File: src/xdist/scheduler/load.py (hunk 2 above)
Comment (amezin) [reply src/xdist/scheduler/load.py:252]: Honestly, I don't know. To me, round-robin distribution it doesn't look optimal even for small collections. But I just want to stay within "trivial changes and reverts".

I see that 09d79ace356ebf14d221cecd3fd8002d075a2034 improved the case with small number of tests because it distributed tests to all workers, not just first `n / 2`. But that doesn't mean you should do round-robin distribution. I'm just keeping the original code where it works better, trying to achieve the result with minimal changes.

Comment (nicoddemus) [reply src/xdist/scheduler/load.py:252]: > But that doesn't mean you should do round-robin distribution.

Actually you got a point... I can't find the original PR with that change.

@RonnyPfannschmidt do you recall why that was change to use round robin?

Comment (amezin) [reply src/xdist/scheduler/load.py]: Maybe it's just simpler to implement than adjusting "sequential" distribution code for the corner case. I've added a comment.

File: src/xdist/scheduler/load.py (hunk 3 above)
Comment (amezin) [reply src/xdist/scheduler/load.py:265]: Done

File: changelog/812.feature (hunk 4 above)
Comment (amezin) [reply changelog/812.feature:1]: Done

File: src/xdist/scheduler/load.py (hunk 2 above)
Comment (amezin) [reply src/xdist/scheduler/load.py]: PR: https://github.com/pytest-dev/pytest-xdist/pull/13 (commit 09d79ace356ebf14d221cecd3fd8002d075a2034 has a link to it)

**Comment by amezin** [comment]
Not sure what's going on with the failed test/what to do about it

File: src/xdist/scheduler/load.py (hunk 2 above)
Comment (nicoddemus) [reply src/xdist/scheduler/load.py]: Thanks! I think one of the reasons is that you have a test module with very slow tests (disregarding fixtures), it is better to distribute that evenly across many workers rather than scheduling most/all of them a single worker. 

I agree with the approach of keeping the changes to a minimal.

File: changelog/812.feature [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+With the old algorithm (before v1.14), if there are enough tests not using
+the fixture, the fixture was created only once.
+
+So restore the old behavior for typical cases where the number of tests is
+much greater than the number of workers (or, strictly speaking, when there
+are at least 2 tests for every node).
+
+In my test suite, where fixtures create Docker containers, this change reduces
+total run time by 10-15%.
```
Comment (nicoddemus) [inline changelog/812.feature:24]:
```suggestion
```

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt would you like to review too? Otherwise I will merge this in the next few days. 👍

**Comment by RonnyPfannschmidt** [comment]
Not sure if I get the bandwidth, it's a nice to have,

File: src/xdist/scheduler/load.py (hunk 2 above)
Comment (amezin) [reply src/xdist/scheduler/load.py]: @nicoddemus Hm... Maybe I chose wrong condition for switching between "old" and "new" algorithms then.

`check_schedule()` has a heuristic for long-running tests. But it will have a chance to rebalance anything only if there are at least 3 tests per node.

Comment (nicoddemus) [reply src/xdist/scheduler/load.py]: Hmm I see. Perhaps let's change this condition to `if len(self.pending) < 3 * len(self.nodes):` then?
